src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/crud-table/crud-table.component.html

```diff
@@ -36,7 +36,7 @@
           <tr *ngFor="let column of meta.detail_columns">
             <td i18n
                 class="bold">{{ column }}</td>
-            <td> {{ expandedRow[column] }} </td>
+            <td><pre>{{ customJSONStringify(expandedRow[column]) }}</pre></td>
```

Comment:
it is generally not advised to use the function calls in templates because of the way angular [change detection is triggered](https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496). 

Maybe we can move this logic or call this from `processMeta` function instead and do the processing if its required and update the `meta.detail_columns` with the newer values and then pass these updated value to the template?

Also this change means that we are passing `expandedRow[column]` through the json stringify even if it is not meant to be. Can we do this conditionally?

Reply:
> it is generally not advised to use the function calls in templates because of the way angular [change detection is triggered](https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496). You can test this out by putting a console log in the function and see how many times its getting called
> 
> Maybe we can move this logic or call this from `processMeta` function instead and do the processing if its required and update the `meta.detail_columns` with the newer values and then pass these updated value to the template?
> 
I can check that surely.. makes sense if its not efficient.
I liked the idea of moving to `processMeta` and I can look into that...
> Also this change means that we are passing `expandedRow[column]` through the json stringify even if it is not meant to be. Can we do this conditionally?
I have changed the JSON.stringify replacer function, If you check the other file, if its a non-string, non-json, it will be returned as is.

The reason for doing this was first I tried `json` pipe which actually not taking care of escape sequences and was rendering them as it is. In a nutshell, it was non formatted json
I tried conditional way as well, and it was hardcoding certain fields or checking type of objects..
The `expandedRow[column]` is something coming as array of objects or strings, or objects.

Reply:
```
 /**
   * Checks strings for JSON value
   * If yes, returns the parsed value
   * If no, returns the original string
   *
   * For non-strings, return as it is
   **/
  customJSONStringify(rowValue: any) {
    return JSON.stringify(
      rowValue,
      (_key: string, value: any) => {
        if (typeof value === 'string') {
          try {
            const parsedValue = JSON.parse(value); 
            return parsedValue; 
          } catch (e) {
            return value; // returns string but non-json object as it is
          }
        }
        return value; // returns non-string values as it is
      },
      2
    );
  }
}
```
I have added a custom replacer function, so only valid values which are meant to be parsed will go through..

Reply:
yup, i agree. probably a generic pipe could be useful later on for us.